### PR TITLE
Beir benchmark

### PR DESCRIPTION
**Pull request by baptistejamin**
[BEIR](https://arxiv.org/pdf/2104.08663) evaluates retrieval and ranking quality across multiple datasets. The Quora dataset is particularly representative of Sonic's search use cases.

## Run the benchmark

```shell
BEIR_REINDEX=1 \
BEIR_STOPWORDS=true \
BEIR_RETAIN_WORD_OBJECTS=1000 \
cargo bench -p sonic-server --bench beir_quora
```

The following results were measured over 10,000 queries.

## Results

| Implementation | Retained objects per term | Stopwords | nDCG@10 | MAP@100 | Recall@100 | Precision@10 |
| :--- | ---: | :---: | ---: | ---: | ---: | ---: |
| Sonic master | 1,000 | Enabled | 0.20432 | 0.18803 | 0.35868 | 0.03088 |
| Sonic master | 1,000 | Disabled | 0.00152 | 0.00135 | 0.00482 | 0.00025 |
| Sonic master | 10,000 | Enabled | 0.44793 | 0.41302 | 0.72547 | 0.06989 |
| Sonic master | 10,000 | Disabled | 0.00346 | 0.00321 | 0.00945 | 0.00052 |
| Sonic master | 100,000 | Enabled | 0.47637 | 0.43948 | 0.75823 | 0.07469 |
| Sonic master | 100,000 | Disabled | 0.03954 | 0.03628 | 0.07077 | 0.00606 |
| Sonic PoC v2 | Unlimited | Disabled | 0.46279 | 0.42657 | 0.73669 | 0.07083 |
| Meilisearch master | — | — | 0.26495 | 0.24946 | 0.34351 | 0.03713 |

For comparison, the BEIR paper reports a BM25 baseline of **0.789 nDCG@10** and **0.973 Recall@100** for Quora.

## Metrics

- **nDCG@10:** Measures how highly relevant documents are ranked in the first 10 results. Higher is better.
- **MAP@100:** Measures ranking quality across the first 100 results, penalizing relevant documents that appear late or are missing.
- **Recall@100:** Measures the fraction of relevant documents found in the first 100 results.
- **Precision@10:** Measures the fraction of the first 10 results that are relevant. A score of `0.03088` represents about 0.31 relevant documents per query.

### CODE REVIEW

File: server/tests/common/spawn_guard.rs [hunk 1]
```diff
@@ -23,7 +47,7 @@ impl std::ops::DerefMut for SpawnGuard {
 
 impl Drop for SpawnGuard {
     fn drop(&mut self) {
-        self.0.kill().unwrap();
-        self.0.wait().unwrap();
+        let _ = self.0.kill();
+        let _ = self.0.wait();
```
Comment (RemiBardon) [inline server/tests/common/spawn_guard.rs]: In such tests, it’s useful to have `unwrap`s. It helps detecting weird edge cases we might not see otherwise. It should never happen, but when it does, it definitely needs a fix. Please revert 🙃

**Review by RemiBardon — CHANGES_REQUESTED** [review]
I overlooked most of the code but I trust you that it’s working as intended (and it’s just benchmark code) so LGTM 🙂

File: server/Cargo.toml [hunk 2]
```diff
@@ -31,7 +31,9 @@ config = { version = "=0.15.25", default-features = false, features = ["toml"] }
 # Benchmarking dependencies.
 arrow-array = { version = "=59.1.0", default-features = false }
 criterion = { version = "=0.8.2", default-features = false, features = ["html_reports", "plotters"] }
+hf-hub = { version = "=0.4.3", default-features = false, features = ["ureq"] }
```
Comment (RemiBardon) [inline server/Cargo.toml]: To fix the CI issue, run `task deps:validate-build-scripts` and check that the build script does nothing malicious. I created this tool so it’s quick and easy to review and validate changes.

Edit: You can install Task via https://taskfile.dev/docs/installation, or you can run `./scripts/validate-dep-build-script-change.sh` directly instead if you don’t want to install yet another tool 😉

**Review by RemiBardon — CHANGES_REQUESTED** [review]

**Comment by RemiBardon** [comment]
Before merging I want to make sure datasets are cached locally, as overlooking the code I can’t say for sure and this was a useful feature of my previous version.

Apart from that, I’ll merge soon.

**Comment by RemiBardon** [comment]
I squashed your commits and made a few touch-ups, I’ll merge after CI passes.

**Review by RemiBardon — APPROVED** [review]
